// database/sqlite.js
const { DataSource } = require('apollo-datasource');
const { Op } = require("sequelize");

const env = require("env_vars");


class SqliteDatasource extends DataSource {

    constructor(db_instance) {
        super()
        this.database = db_instance.db
        this.models = db_instance.models
        env.logger.setDebugLevels("database")
    }

    _run_complex_query(table, query) {
        env.logger.debug(1, "Querying", table, "for", JSON.stringify(query))
        env.logger.resetDebugLevels()

        let find = { where: query }
        return this.models[table].findAll(find).then(
            response => { return response })
    }

    findAll(table) {
        env.logger.debug(1, "Selecting all rows for ", table)
        env.logger.resetDebugLevels()
        return this.models[table].findAll()
    }

    get_string_query(args) {
        let query = {}
        Object.keys(args).forEach((arg) => {
            if (typeof args[arg] != 'string')
                throw TypeError(`'${args[arg]}' is not a string`)

            let values = args[arg].split(' ').join(',').split(',') // split on spaces and commas
            let and_query = []
            values.forEach((value) => {
                and_query.push({[Op.like]: `%${value}%`})
            })
            query[arg] = { [Op.and]: and_query }
            console.log(query)
        })
        return query
    }

    get_betwen(table, column, values) {
        let query = {[Op.between]: values}
        return this._run_complex_query(table, column, query)
    }

    get_over(table, column, value) {
        let query = {[Op.gt]: value}
        return this._run_complex_query(table, column, query)
    }

    get_under(table, column, value) {
        let query = {
            [Op.or]: {
              [Op.lt]: value,
              [Op.eq]: null
            }
        }
        return this._run_complex_query(table, column, query)
    }

    _get_by(table, column, value) {
        let query = { where: {} }
        query.where[column] = value
        return this.models[table].findOne(query)
    }

    get_by_id(table, id) {
        return this._get_by(table, "id", id)
    }

    get_by_name(table, name) {
        return this._get_by(table, "name", name)
    }
}

module.exports = SqliteDatasource
